qiskit_machine_learning/kernels/statevector_kernel.py

```diff
@@ -0,0 +1,191 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Fidelity Statevector Kernel"""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.quantum_info import Statevector
+
+from .base_kernel import BaseKernel
+
+
+class FidelityStatevectorKernel(BaseKernel):
+    r"""
+    A reference implementation of the quantum kernel interface limited to classically simulated
+    statevectors.
+
+    The general task of machine learning is to find and study patterns in data. For many
+    algorithms, the datapoints are better understood in a higher dimensional feature space,
+    through the use of a kernel function:
+
+    .. math::
+
+        K(x, y) = \langle f(x), f(y)\rangle.
+
+    Here :math:`K` is the kernel function, :math:`x`, :math:`y` are :math:`n` dimensional inputs.
+    :math:`f` is a map from :math:`n`-dimension to :math:`m`-dimension space. :math:`\langle x, y
+    \rangle` denotes the dot product. Usually :math:`m` is much larger than :math:`n`.
+
+    The quantum kernel algorithm calculates a kernel matrix, given datapoints :math:`x` and
+    :math:`y` and feature map :math:`f`, all of :math:`n` dimension. This kernel matrix can then be
+    used in classical machine learning algorithms such as support vector classification, spectral
+    clustering or ridge regression.
+
+    Here, the kernel function is defined as the overlap of two simulated quantum statevectors
+    produced by a parametrized quantum circuit (called feature map):
+
+    .. math::
+
+        K(x,y) = |\langle \phi(x) | \phi(y) \rangle|^2.
+
+    In this implementation, :math:`\phi` is represented by a ``Statevector.data`` array,
+    thus the kernel function is given simply by
+
+    .. math::
+
+        K(x,y) = (\phi(x)^\dagger \phi(y))^2.
```

Comment:
This misses the absolute value
```suggestion
        K(x,y) = |\phi(x)^\dagger \phi(y)|^2.
```

Reply:
Naive question, since this is squared... does the absolute value serve any purpose?

Reply:
the overlap can be complex in general :)

Reply:
E.g. `\phi(x) = [0, 1]` and `\phi(y) = [0, i]` would evaluate to `-i` and then squaring would give a negative kernel entry 😄